benchmarks/brownian_tree_times.py

```diff
@@ -0,0 +1,191 @@
+            )
+
+    def _evaluate(self, τ: RealScalarLike) -> PyTree[Array]:
+        map_func = lambda key, struct: self._evaluate_leaf(key, τ, struct)
+        return jtu.tree_map(map_func, self.key, self.shape)
+
+    def _brownian_bridge(self, s, t, u, w_s, w_u, key, shape, dtype):
+        mean = w_s + (w_u - w_s) * ((t - s) / (u - s))
+        var = (u - t) * (t - s) / (u - s)
+        std = jnp.sqrt(var)
+        return mean + std * jr.normal(key, shape, dtype)
+
+    def _evaluate_leaf(
+        self,
+        key,
+        τ: RealScalarLike,
+        struct: jax.ShapeDtypeStruct,
+    ) -> Array:
+        shape, dtype = struct.shape, struct.dtype
+
+        cond = self.t0 < self.t1
+        t0 = jnp.where(cond, self.t0, self.t1).astype(dtype)
+        t1 = jnp.where(cond, self.t1, self.t0).astype(dtype)
+
+        t0 = eqxi.error_if(
+            t0,
+            τ < t0,
+            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
+        )
+        t1 = eqxi.error_if(
+            t1,
+            τ > t1,
+            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
+        )
+        τ = jnp.clip(τ, t0, t1).astype(dtype)
+
+        key, init_key = jr.split(key, 2)
+        thalf = t0 + 0.5 * (t1 - t0)
+        w_t1 = jr.normal(init_key, shape, dtype) * jnp.sqrt(t1 - t0)
+        w_thalf = self._brownian_bridge(t0, thalf, t1, 0, w_t1, key, shape, dtype)
+        init_state = _State(
+            s=t0,
+            t=thalf,
+            u=t1,
+            w_s=jnp.zeros_like(w_t1),
+            w_t=w_thalf,
+            w_u=w_t1,
+            key=key,
+        )
+
+        def _cond_fun(_state):
+            return (_state.u - _state.s) > self.tol
+
+        def _body_fun(_state):
+            _key1, _key2 = jr.split(_state.key, 2)
+            _cond = τ > _state.t
+            _s = jnp.where(_cond, _state.t, _state.s)
+            _u = jnp.where(_cond, _state.u, _state.t)
+            _w_s = jnp.where(_cond, _state.w_t, _state.w_s)
+            _w_u = jnp.where(_cond, _state.w_u, _state.w_t)
+            _key = jnp.where(_cond, _key1, _key2)
+            _t = _s + 0.5 * (_u - _s)
+            _w_t = self._brownian_bridge(_s, _t, _u, _w_s, _w_u, _key, shape, dtype)
+            return _State(s=_s, t=_t, u=_u, w_s=_w_s, w_t=_w_t, w_u=_w_u, key=_key)
+
+        final_state = lax.while_loop(_cond_fun, _body_fun, init_state)
+
+        s = final_state.s
+        u = final_state.u
+        w_s = final_state.w_s
+        w_t = final_state.w_t
+        w_u = final_state.w_u
+        rescaled_τ = (τ - s) / (u - s)
+
+        A = jnp.array([[2, -4, 2], [-3, 4, -1], [1, 0, 0]])
+        coeffs = jnp.tensordot(A, jnp.stack([w_s, w_t, w_u]), axes=1)
+        return jnp.polyval(coeffs, rescaled_τ)
+
+
+def time_tree(tree_cls, tol, levy_area):
+    tree = tree_cls(t0=t0, t1=t1, tol=tol, shape=(100,), key=key, levy_area=levy_area)
+
+    def f():
+        return jax.block_until_ready(vec_eval(tree, ts))
```

Comment:
I think there should be a `jax.jit` here as `vec_eval` involves a `jax.vmap`, which is a JAX operation happening outside of vmap.

Also, for timing benchmarks like this it is better to do many repeats and take a minimum. (Noise can only increase things from the best possible time, so you want a minimum, not a mean.)

Thus I think you want something like:
```python
@jax.jit
def run(ts):
    return jax.vmap(lambda _t: tree.evaluate(_t, use_levy=True))(_ts)

return min(timeit.repeat(lambda: jax.block_until_ready(run(ts)), number=1, repeat=100))
```

Reply:
Oh I see, so any JAX operation should be wrapped inside a jit?

Reply:
Yup! At least, if you want them to fast.

So for typical neural network training, I often won't bother putting a JIT when `__init__`ialising a model (which only happens once), but will definitely put one around the entirety of the forward pass.